### PR TITLE
Improve Encoding support in SyntaxTrees and Documents

### LINKED ISSUE #934: Bug in TryApplyChanges causes it to rewrite all line endings
The bottom of this issue has a complete code snippet that demonstrates the problem.  It's as minimal as I could make it but in summary it does the following:

- Opens up the CodeAnalysis project.
- Defines a symbol that isn't used anywhere in the code in the parse options.
- Runs the formatter on every file.
- Undefines the symbol. 
- Applies the changes.

The apply changes portion will cause every document in the project to be rewritten with different line endings.  This happens even if the formatter operation is a complete no-op.  

The problem appears to be the changing and resetting of the parse options.  If the same sample is run without changing the preprocessor symbols (comment out the line labeled CRITICAL LINE) then the sample behaves as expected.  

This bug came up as a result of running the code formatter on Roslyn.  It caused Git to believe every file had a binary difference.  

``` csharp
var projectPath = @"e:\dd\ros\Open\src\Compilers\Core\Portable\CodeAnalysis.csproj";
var workspace = MSBuildWorkspace.Create();
workspace.LoadMetadataForReferencedProjects = true;
var project = await workspace.OpenProjectAsync(projectPath, CancellationToken.None);
var originalOptions = (CSharpParseOptions)project.ParseOptions;
var originalSymbols = originalOptions.PreprocessorSymbolNames;
var newSymbols = originalSymbols.Concat(new[] { "TEST_SYMBOL" });

// CRITICAL LINE 
project = project.WithParseOptions(originalOptions.WithPreprocessorSymbols(newSymbols));

foreach (var doc in project.Documents)
{
    var syntaxNode = await doc.GetSyntaxRootAsync(CancellationToken.None);
    if (syntaxNode == null)
    {
        continue;
    }

    var newNode = Formatter.Format(syntaxNode, workspace);
    project = doc.WithSyntaxRoot(newNode).Project;
}

project = project.WithParseOptions(originalOptions);
if (workspace.TryApplyChanges(project.Solution))
{
    Console.WriteLine("Succeeded");
}
```

### PR DESCRIPTION
**Pull request by mattwar**
Fixes #934 

This changes adds better tracking of text encoding during normal document edits, increasing the probability that the SourceText and SyntaxTree have a non-null Encoding property.

Adds public Encoding property to SyntaxTree.

### CODE REVIEW

**Comment by mattwar** [comment]
@pilchie, @tmat, @pharring, @jaredpar, @jasonmalinowski  please review.

**Comment by mattwar** [comment]
note: an alternative to calling TryGetText in DocumentState.UpdateRoot would be to call GetText and block, forcing the text to load if it hadn't already, thus always obtaining the correct encoding, and never needing to set it to null. However, this still wouldn't guarantee that the encoding was never null, just that it always was the same as the original SourceText, which could have been null to begin with.

**Comment by mattwar** [comment]
@tmat opportunistically cleaning up bad style.  We chose long ago (years) no longer to use Opt suffix.

**Comment by tmat** [comment]
For public APIs. Not for internal APIs. This style is not bad. On contrary it saves a lot of effort trying to figure out whether or not null is expected value. Until we have a language feature to mark nullable reference types we should keep using Opt suffix internally.

**Comment by Pilchie** [comment]
It's curious that this now allows a `SyntaxTree` and it's `SourceText` to have different values for Encoding.

Otherwise :+1:

**Comment by mattwar** [comment]
undid opt prefix removal, since this pull request is not the place for debating coding style.

**Comment by tmat** [comment]
Thanks :)

**Comment by mattwar** [comment]
@pilchie, technically yes, but in practice, no.. one of the values is generally used to produce the other in the actual implementations.

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs [hunk 1]
```diff
@@ -368,9 +368,28 @@ internal DocumentState UpdateTree(SyntaxNode newRoot, PreservationMode mode)
...
+            else if (this.TryGetText(out priorText))
+            {
+                encoding = priorText.Encoding;
+            }
+            else
+            {
+                // the existing encoding was never observed so is unknown.
+                encoding = null;
+            }
+
```
Comment (pharring) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: Refactor into a helper method?

File: src/Workspaces/CoreTest/WorkspaceTests/MSBuildWorkspaceTests.cs [hunk 2]
```diff
@@ -2376,19 +2376,40 @@ class C { }";
...
+            // change doc to have no encoding, still succeeds at writing to disk with old encoding
+            var root = document.GetSyntaxRootAsync().Result;
+            var noEncodingDoc = document.WithText(SourceText.From(text.ToString(), encoding: null));
+            Assert.Null(noEncodingDoc.GetTextAsync().Result.Encoding);
+            
+            // apply changes (this writes the changed document)
+            var noEncodingSolution = noEncodingDoc.Project.Solution;
+            Assert.True(noEncodingSolution.Workspace.TryApplyChanges(noEncodingSolution));
+
+            // prove the written document still has the same encoding
```
Comment (pharring) [inline src/Workspaces/CoreTest/WorkspaceTests/MSBuildWorkspaceTests.cs]: Nice set of tests.

**Comment by pharring** [comment]
:+1:

**Pull request merged**